File: agileforge-backend/src/main/java/com/agileforge/application/service/BoardService.java
package com.agileforge.application.service;

import com.agileforge.domain.exception.EntityNotFoundException;
import com.agileforge.domain.model.*;
import com.agileforge.domain.port.out.BoardColumnRepositoryPort;
import com.agileforge.domain.port.out.ProjectRepositoryPort;
import com.agileforge.domain.port.out.SprintRepositoryPort;
import com.agileforge.domain.port.out.TicketRepositoryPort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;

@Service
@Transactional
public class BoardService {

    private final BoardColumnRepositoryPort columnRepository;
    private final TicketRepositoryPort ticketRepository;
    private final SprintRepositoryPort sprintRepository;
    private final ProjectRepositoryPort projectRepository;

    public BoardService(BoardColumnRepositoryPort columnRepository,
                        TicketRepositoryPort ticketRepository,
                        SprintRepositoryPort sprintRepository,
                        ProjectRepositoryPort projectRepository) {
        this.columnRepository = columnRepository;
        this.ticketRepository = ticketRepository;
        this.sprintRepository = sprintRepository;
        this.projectRepository = projectRepository;
    }

    @Transactional(readOnly = true)
    public BoardData getBoard(UUID projectId) {
        Project project = projectRepository.findById(projectId)
                .orElseThrow(() -> new EntityNotFoundException("Project", projectId));

        List<BoardColumn> columns = columnRepository.findByProjectIdOrderByPosition(projectId);

        // If no columns configured, create default ones
        if (columns.isEmpty()) {
            columns = createDefaultColumns(projectId);
        }

        Sprint activeSprint = sprintRepository.findActiveByProjectId(projectId).orElse(null);

        List<Ticket> tickets;
        if (activeSprint != null) {
            tickets = ticketRepository.findBySprintId(activeSprint.getId());
        } else {
            tickets = ticketRepository.findByProjectId(projectId).stream()
                    .filter(t -> !t.isDone())
                    .toList();
        }

        return new BoardData(project, columns, tickets, activeSprint);
    }

    public BoardColumn addColumn(UUID projectId, String name, TicketStatus mappedStatus, int position, Integer wipLimit) {
        BoardColumn column = new BoardColumn(projectId, name, mappedStatus, position);
        column.setWipLimit(wipLimit);
        return columnRepository.save(column);
    }

    public void removeColumn(UUID columnId) {
        columnRepository.delete(columnId);
    }

    public void moveTicket(UUID ticketId, TicketStatus newStatus) {
        Ticket ticket = ticketRepository.findById(ticketId)
                .orElseThrow(() -> new EntityNotFoundException("Ticket", ticketId));
        ticket.setStatus(newStatus);
        ticketRepository.save(ticket);
    }

    @Transactional(readOnly = true)
    public List<Ticket> getBacklog(UUID projectId) {
        return ticketRepository.findByProjectIdAndStatus(projectId, TicketStatus.BACKLOG);
    }

    private List<BoardColumn> createDefaultColumns(UUID projectId) {
        BoardColumn[] defaults = {
                new BoardColumn(projectId, "Backlog", TicketStatus.BACKLOG, 0),
                new BoardColumn(projectId, "To Do", TicketStatus.TODO, 1),
                new BoardColumn(projectId, "In Progress", TicketStatus.IN_PROGRESS, 2),
                new BoardColumn(projectId, "Code Review", TicketStatus.CODE_REVIEW, 3),
                new BoardColumn(projectId, "QA", TicketStatus.QA, 4),
                new BoardColumn(projectId, "Done", TicketStatus.DONE, 5)
        };

        for (BoardColumn col : defaults) {
            columnRepository.save(col);
        }

        return columnRepository.findByProjectIdOrderByPosition(projectId);
    }

    public record BoardData(Project project, List<BoardColumn> columns, List<Ticket> tickets, Sprint activeSprint) {}
}

File: agileforge-frontend/src/app/core/services/board.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Board } from '../models/board.model';
import { Ticket } from '../models/ticket.model';
import { environment } from '../../../environments/environment';

@Injectable({ providedIn: 'root' })
export class BoardService {
  private readonly apiUrl = `${environment.apiUrl}/board`;

  constructor(private http: HttpClient) {}

  getBoard(projectId: string): Observable<Board> {
    return this.http.get<Board>(`${this.apiUrl}/project/${projectId}`);
  }

  moveTicket(ticketId: string, status: string): Observable<void> {
    return this.http.patch<void>(`${this.apiUrl}/tickets/${ticketId}/move?status=${status}`, {});
  }

  addColumn(projectId: string, name: string, mappedStatus: string, position: number, wipLimit?: number): Observable<any> {
    return this.http.post(`${this.apiUrl}/project/${projectId}/columns`, { name, mappedStatus, position, wipLimit });
  }

  removeColumn(columnId: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/columns/${columnId}`);
  }

  getBacklog(projectId: string): Observable<Ticket[]> {
    return this.http.get<Ticket[]>(`${this.apiUrl}/project/${projectId}/backlog`);
  }
}
